### PR TITLE
refactor: Move component output into component file

### PR DESCRIPTION
**Pull request by tofumatt**
## Summary

Addresses issue #427.

This moves the output closer to the render method of this component; at least in the same file.

Unfortunately, because of how the data layer is de-coupled from state and the way the components are modifying each other's state, this was complicated to untangle and ultimately resulted in better but still not great code in my view. The existing code relied on many magic constants which are still intact and I'm not clear where they were decided.

Overall I think this is better but is a clear demonstration of the limitations of refactoring existing code before we have solved more fundamental issues like asynchronous actions, data fetching, and state management.

This also took a lot of time because the existing code paths were quite hard to follow and testing of this component, both in E2E and manually, took a lot of time. Each test run, just for this component, took `232s` on my i7 MacBook Pro.

The high amounts of branching in this component (based on `accountStatus`) makes it an excellent candidate for `jest` unit testing/snapshot testing, if there was much doubt about it versus E2E 😄. (See: #437)

## Relevant technical choices

This one is finally ready for code review. The code has been moved into the component file and the code, I think, simplified, but the state management in the app/module leaves a lot to be desired so I didn't mess with that too much. I took the opportunity to use some fancy new hooks in the code to further reduce boilerplate.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/util/react-features.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/**
```
Comment (tofumatt) [inline assets/js/util/react-features.js]: If you're cool with this approach, @aaemnnosttv, I can update existing parts of the codebase to use this module. I think we're much better off importing these things from one place where we control the import path than the approach used elsewhere, eg:

https://github.com/google/site-kit-wp/blob/767c01942c2909e69d1a4b03504ab3d270f22552/assets/js/googlesitekit-module.js#L42-L50

Comment (aaemnnosttv) [reply assets/js/util/react-features.js]: Yeah, good idea! I definitely prefer a single place to import from 👍

File: assets/js/modules/tagmanager/setup.js [hunk 2]
```diff
@@ -379,12 +379,16 @@ class TagmanagerSetup extends Component {
...
 						outlined
 					>
-						{ accounts.map( ( account ) =>
-							<Option
-								key={ account.accountId }
-								value={ account.accountId }>
-								{ account.name }
-							</Option> ) }
+						{ accounts.map( ( account ) => {
+							window.console.log( 'account', account );
```
Comment (aaemnnosttv) [inline assets/js/modules/tagmanager/setup.js]:
```suggestion
```

File: assets/js/modules/tagmanager/setup.js [hunk 3]
```diff
@@ -396,16 +400,20 @@ class TagmanagerSetup extends Component {
...
-									key={ container.containerId }
-									value={ container.publicId }>
-									{
-										0 === container.publicId ?
-											__( 'Set up a new container', 'google-site-kit' ) :
-											container.publicId
-									}
-								</Option> ) }
+							{ containers.map( ( container ) => {
+								window.console.log( 'container', container );
```
Comment (aaemnnosttv) [inline assets/js/modules/tagmanager/setup.js]:
```suggestion
```

File: tests/e2e/specs/modules/adsense/setup-new-user.test.js [hunk 4]
```diff
@@ -138,7 +140,9 @@ describe( 'setting up the AdSense module', () => {
 
 		await proceedToAdsenseSetup();
 
-		await expect( page ).toMatchElement( '.googlesitekit-setup-module__title', { text: /We’re getting your site ready for ads/i } );
+		await page.waitForSelector( '.googlesitekit-setup-module--adsense' );
+
+		await expect( page ).toMatchElement( '.googlesitekit-heading-4.googlesitekit-setup-module__title', { text: /We’re getting your site ready for ads/i } );
```
Comment (aaemnnosttv) [inline tests/e2e/specs/modules/adsense/setup-new-user.test.js]: Is the `.googlesitekit-heading-4` class necessary here? I'd prefer to not add presentational classes to the selector if possible.

File: tests/e2e/specs/modules/adsense/setup-new-user.test.js [hunk 5]
```diff
@@ -196,6 +200,8 @@ describe( 'setting up the AdSense module', () => {
 
 		await proceedToAdsenseSetup();
 
+		await page.waitForSelector( '.googlesitekit-setup-module--adsense' );
```
Comment (aaemnnosttv) [inline tests/e2e/specs/modules/adsense/setup-new-user.test.js]: We could move this to `proceedToAdsenseSetup` if it's needed for every test?

File: assets/vendor/react-dom.development.js [hunk 6]
```diff
@@ -19848,4 +19848,4 @@ var reactDom = ReactDOM$3.default || ReactDOM$3;
 
 return reactDom;
 
-})));
+})));
```
Comment (aaemnnosttv) [inline assets/vendor/react-dom.development.js]: Why would this file have changed?

File: tests/e2e/specs/modules/adsense/setup-new-user.test.js (hunk 5 above)
Comment (tofumatt) [reply tests/e2e/specs/modules/adsense/setup-new-user.test.js]: Actually, it was already in there… so I'm gonna remove it. Might've slipped in during debug, cheers!

File: assets/vendor/react-dom.development.js (hunk 6 above)
Comment (tofumatt) [reply assets/vendor/react-dom.development.js]: I was trying to debug errors and needed to load a particular version of `react-dom`… looks like I must've had it open when restoring and my editor added a newline EOF. I'll revert it, thanks for catching that (the woes of `git add .` 😓)

File: tests/e2e/specs/modules/adsense/setup-new-user.test.js (hunk 4 above)
Comment (tofumatt) [reply tests/e2e/specs/modules/adsense/setup-new-user.test.js]: Oh, good call. It wasn't; I've removed it 👍

**Comment by tofumatt** [comment]
I tweaked the React feature checks and things were passing locally.

Oddly, `WP_VERSION=4.9.10 GUTENBERG_VERSION=4.9.0 npm run env:start && npm run test:e2e` runs fine on my Mac, but I'm still seeing Travis failures for those tasks. I just saw one run fail on a weird Docker EOF error, which seems totally unrelated, so I'm gonna re-start for a few failures in case it's a weird intermittent thing, which I know happens on Travis with E2E tests…

**Comment by tofumatt** [comment]
Seems like there is a greater asset-loading issue I’m running into here. In the case of WordPress pre-5.0, the `wp.element` global isn't available and we fallback to the React and ReactDOM assets in the `vendor` folder. But the version there isn't actually up-to-date, despite us requesting a newer version we're getting the file in the `assets/vendor/` folder which is pre-hooks: https://github.com/google/site-kit-wp/blob/develop/assets/vendor/react-dom.development.js#L1-L8

I'm trying to fix this now; it's possible we should just upgrade this file, but really it should match the version we specify in `package.json`—otherwise, it's _very_ confusing.

**Comment by tofumatt** [comment]
Ah, frustratingly, the issues here stem from Gutenberg bundling its own version of `react` and `react-dom`. It's not clear to me _why_ we test 4.9 _with_ the Gutenberg plugin (especially a very old, unsupported version)—I think we need to stop doing that.

Running a non-current version of Gutenberg will cause issues with anything using a non-bundled version of React—which we do because we don't bundle React with our webpack bundle.

Removing `GUTENBERG_VERSION=4.9.0` and loading newer versions of React to the `vendor/` folder fixes this issue. I'm going to tidy up my vendor-handling code and submit a patch to this PR tomorrow 😄

**Comment by tofumatt** [comment]
This now works, but as mentioned: requires we stop specifying an old version of Gutenberg as part of the testing environment. Running Gutenberg 4.9 doesn't make much sense, as it's [quite out-of-date](https://github.com/WordPress/gutenberg/releases/tag/v4.9.0), and shouldn't be considered supported.

**Comment by felixarntz** [comment]
@tofumatt 
> Running Gutenberg 4.9 doesn't make much sense, as it's quite out-of-date, and shouldn't be considered supported.

We previously aligned on a product requirement to support WordPress back to version 4.7, and it is crucial to have a testing scenario with pre-Gutenberg WordPress and the latest Gutenberg version that supports that WordPress version (which should be 4.9). We cannot just drop this, even though it might be a pain to support.

We can discuss this at a broader scope, and limiting the backward-compatibility pre-release to something more reasonable might make sense. But, at the same time, this limits the plugin to being used by a smaller audience, so product leadership here has the final say.

**Comment by tofumatt** [comment]
The tests failing on Travis is a bit baffling to me as I've got them passing locally on WP < 5.0 environments.

The current loading of resources seems to be all over the map. When Gutenberg is not loaded, the vendored assets are loaded (I've updated them in this branch). When Gutenberg _is_ loaded, it seems to use a bundled version of React that Gutenberg loads. This bundled version (for Gutenberg 4.9) is quite out-of-date (16.6) so lacks newer features.

I think we should be using our own version of React in this case if we aren't using our own version of the `@wordpress/` modules, though maybe _that's_ what we should be doing. Right now the dependency story is quite confusing, and while in this PR I'm just trying to use fun hooks, going forward there could be security updates or other things we want in future versions of React we should be able to integrate.

The thing about old versions of WordPress I understand: supporting pre-5.0 WordPress makes sense. But why _include_ Gutenberg in that pre-5.0 story? That version of Gutenberg doesn't even—to my knowledge—receive security updates. Users on pre-5.0 who want Gutenberg would upgrade to 5.x, I'd think. The users _not_ on 5.0 are likely staying behind explicitly to avoid Gutenberg, so including an older version that also loads an older React seems… odd to me.

At any rate, I'm having a quick look at the failures here but I'm guessing we'll just need to back out the hooks code here or start including our own versions of these scripts. 🤷‍♂

**Comment by tofumatt** [comment]
On first inspection, it seems that the tests fail because `wp` globals aren't available. This is odd, as they should be installed as a `devDependency` I thought.

**Comment by tofumatt** [comment]
I'm going to switch over to looking at some IBs and other things this afternoon, but in my testing: bundling our own version of React and ReactDOM gets this working in WP 4.7 and WP 4.9, but **not** WP 4.9+Gutenberg 4.9—possibly because of the `wp.` modules. I think it's best we have control over what versions of critical libraries like React we load _regardless_ of this particular issue, so I think it's worth bundling it here.

I know the Jest branch also switches our includes of `wp.` modules over to using `@wordpress/` as well, and it's possible even those we should be using our own version for; I know it increases bundle size, but the results otherwise will be quite unpredictable...

**Comment by tofumatt** [comment]
We chatted about this in a stand-up, and for now the solution will be to not rely on newer features (like hooks).

There is still the deeper issue of not knowing exactly what versions of third-party libraries are available, but we'll need to discuss that further separately.

**Comment by tofumatt** [comment]
This one's finally passing in all environments; for now I've reverted to using class components. Should be ready for a code review now.

**Comment by aaemnnosttv** [comment]
> Users on pre-5.0 who want Gutenberg would upgrade to 5.x, I'd think. The users not on 5.0 are likely staying behind explicitly to avoid Gutenberg, so including an older version that also loads an older React seems odd to me.

That's a good point which is worth discussing more. Let's keep track of these for a larger discussion around minimum supported versions.

File: assets/js/modules/adsense/dashboard/adsense-module-status.js [hunk 7]
```diff
@@ -24,131 +24,303 @@ import SvgIcon from 'GoogleUtil/svg-icon';
...
+			ctaLink = 'https://google.com/admob';
+			break;
+		case 'no-account':
+			statusHeadline = __( 'Create your AdSense account', 'google-site-kit' );
+			statusMessage = __( 'Site Kit will place AdSense code on every page across your site. This means Google will automatically place ads for you in all the best places.', 'google-site-kit' );
+			profile = true;
+			ctaLinkText = __( 'Create AdSense Account', 'google-site-kit' );
+			ctaLink = signupURL;
+			footerText = __( 'Already have an AdSense account?', 'google-site-kit' );
+			footerAppendedText = __( 'to connect to it', 'google-site-kit' );
```
Comment (aaemnnosttv) [inline assets/js/modules/adsense/dashboard/adsense-module-status.js]: This is unrelated to the refactoring here, but this is improper use of translation strings as these should never be concatenated/appended.

I'll create a new issue for this.

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

File: assets/js/modules/adsense/dashboard/adsense-module-status.js (hunk 7 above)
Comment (tofumatt) [reply assets/js/modules/adsense/dashboard/adsense-module-status.js]: Hah, good catch. There are so many strings and things happening in this PR that I lost sight of some of the smaller things. Cheers 👍

File: assets/js/modules/adsense/dashboard/adsense-module-status.js [hunk 8]
```diff
@@ -24,131 +24,304 @@ import SvgIcon from 'GoogleUtil/svg-icon';
...
+const { Component } = wp.element;
 const { __ } = wp.i18n;
 
-class AdSenseModuleStatus extends Component {
+const propsFromAccountStatus = ( accountStatus, existingTag ) => {
+	/**
+	 * Defines the account status variables.
+	 */
+	let accountTagMatch = false;
+	const clientId = false;
```
Comment (felixarntz) [inline assets/js/modules/adsense/dashboard/adsense-module-status.js]: What's this about? I'm not sure why we would use a `const` set to `false` here. What is this even used for then?

File: assets/js/modules/adsense/dashboard/adsense-module-status.js [hunk 9]
```diff
@@ -24,131 +24,304 @@ import SvgIcon from 'GoogleUtil/svg-icon';
...
+		const status = await getAdSenseAccountStatus( existingTag, setLoadingMessage );
 
-		return (
-			<AdSenseSetupInstructions
-				{ ...this.state }
-				continueSetup={ this.continueSetup }
-				goBack={ this.goBack }
-			/>
-		);
+		this.setState( { accountStatus: status.accountStatus } );
```
Comment (felixarntz) [inline assets/js/modules/adsense/dashboard/adsense-module-status.js]: We don't use the `clientID` returned from the function at all - that seems problematic. Or otherwise, why is it returned at all then?

File: assets/js/modules/adsense/util.js [hunk 10]
```diff
@@ -62,53 +58,52 @@ export function reduceAdSenseData( rows ) {
...
+
+				if ( ! matches || 0 === matches.length ) {
+					accountStatus = 'account-pending-review';
+					sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_account_pending', 'accountPendingReview status account-pending-review' );
+				} else {
+					id = matches[ 0 ].id;
+					sendAnalyticsTrackingEvent( 'adsense_setup', 'adsense_account_detected' );
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply assets/js/modules/adsense/util.js]: I don't think anything new was added, this logic is copied over from existing code, see: https://github.com/google/site-kit-wp/pull/616/files#diff-f4aae6bc04d54dc63925354a972a4186L194

File: assets/js/modules/adsense/dashboard/adsense-module-status.js (hunk 8 above)
Comment (tofumatt) [reply assets/js/modules/adsense/dashboard/adsense-module-status.js]: I think some of these variables were previously unused, and I think this one stuck around. I'll remove it.

File: assets/js/modules/adsense/dashboard/adsense-module-status.js (hunk 9 above)
Comment (tofumatt) [reply assets/js/modules/adsense/dashboard/adsense-module-status.js]: The previous functions were returning a _lot_ of stuff without much documentation. I removed a lot but missed this one; it's since been removed. 👍

**Comment by tofumatt** [comment]
I've merged the latest `develop` here and it applies cleanly. I think I've addressed the concerns you mentioned above, and all tests now pass. This should be good to go 😄
